src/Analyzers/CSharp/Analyzers/ConvertNameOf/CSharpConvertNameOfDiagnosticAnalyzer.cs

```diff
@@ -52,24 +53,24 @@ private void AnalyzeSyntaxAction(SyntaxNodeAnalysisContext syntaxContext)
 
             // TODO: Check for compiler errors on the declaration
 
-            var parent = node.Parent;
-
             // We know that it is a typeof() instance, but we only want to offer the fix if it is a .Name access
+            var parent = node.Parent;
             if (!(node.Parent.IsKind(SyntaxKind.SimpleMemberAccessExpression) && parent.IsNameMemberAccess()))
             {
                 return;
             }
 
-            // TODO: if argument is primitive cases
-
-            //TODO: if argument is generic
-
+            // Analyze the argument to determine if argument is generic or a primitive 
+            if (ArgumentIsGeneric(typeofOp.Operation) || ArgumentIsPrimitive(typeofOp.Operation))
+            {
+                return;
+            }
 
             // Current case can be effectively changed to a nameof instance so report a diagnostic
             var location = Location.Create(syntaxTree, parent.Span);
             var additionalLocations = ImmutableArray.Create(node.GetLocation());
 
-            syntaxContext.ReportDiagnostic(DiagnosticHelper.Create(
+            typeofOp.ReportDiagnostic(DiagnosticHelper.Create(
                      Descriptor,
                      location,
                      ReportDiagnostic.Hidden,
```

Comment:
Unless you plan to add a code style option for this analyzer, you can directly invoke `Diagnostic.Create(Descriptor, location)` here.